cpp/daal/src/algorithms/tsne/tsne_gradient_descent_impl.i

```diff
@@ -0,0 +1,1050 @@
+
+        DataType xMin = posx[iStart];
+        DataType xMax = posx[iStart];
+        DataType yMin = posy[iStart];
+        DataType yMax = posy[iStart];
+        for (IdxType i = iStart + 1; i < iEnd; ++i)
+        {
+            xMin = services::internal::min<cpu, DataType>(xMin, posx[i]);
+            xMax = services::internal::max<cpu, DataType>(xMax, posx[i]);
+            yMin = services::internal::min<cpu, DataType>(yMin, posy[i]);
+            yMax = services::internal::max<cpu, DataType>(yMax, posy[i]);
+        }
+        xMinValues[iBlock] = xMin;
+        xMaxValues[iBlock] = xMax;
+        yMinValues[iBlock] = yMin;
+        yMaxValues[iBlock] = yMax;
+    });
+    DataType xMin = xMinValues[0];
+    DataType xMax = xMaxValues[0];
+    DataType yMin = yMinValues[0];
+    DataType yMax = yMaxValues[0];
+
+    for (IdxType i = 1; i < nBlocks; ++i)
+    {
+        xMin = services::internal::min<cpu, DataType>(xMin, xMinValues[i]);
+        xMax = services::internal::max<cpu, DataType>(xMax, xMaxValues[i]);
+        yMin = services::internal::min<cpu, DataType>(yMin, yMinValues[i]);
+        yMax = services::internal::max<cpu, DataType>(yMax, yMaxValues[i]);
+    }
+
+    //scale the maximum to get all points strictly in the bounding box
+    if (xMax >= 0.)
+        xMax = services::internal::max<cpu, DataType>(xMax * (1. + 1e-3), xMax + 1e-3);
+    else
+        xMax = services::internal::max<cpu, DataType>(xMax * (1. - 1e-3), xMax + 1e-3);
+    if (yMax >= 0.)
+        yMax = services::internal::max<cpu, DataType>(yMax * (1. + 1e-3), yMax + 1e-3);
+    else
+        yMax = services::internal::max<cpu, DataType>(yMax * (1. - 1e-3), yMax + 1e-3);
+
+    //save results
+    //radius       = services::internal::max<cpu, DataType>(xMax - xMin, yMax - yMin) * 0.5 + 1e-5;
+    radius       = services::internal::max<cpu, DataType>(xMax - xMin, yMax - yMin) * 0.5;
+    posx[nNodes] = (xMin + xMax) * 0.5;
+    posy[nNodes] = (yMin + yMax) * 0.5;
+
+    //free memory
+    services::internal::service_free<DataType, cpu>(xMinValues);
+    services::internal::service_free<DataType, cpu>(xMaxValues);
+    services::internal::service_free<DataType, cpu>(yMinValues);
+    services::internal::service_free<DataType, cpu>(yMaxValues);
+
+    return services::Status();
+}
+
+template <typename IdxType, typename DataType, daal::CpuType cpu>
+services::Status qTreeBuildingKernelImpl(IdxType * child, const DataType * posx, const DataType * posy, const IdxType nNodes, const IdxType N,
+                                         IdxType & maxDepth, IdxType & bottom, const DataType & radius)
+{
+    // internal variables
+    IdxType j      = 0;
+    IdxType depth  = 0;
+    IdxType ch     = 0;
+    IdxType n      = 0;
+    IdxType locked = 0;
+    IdxType patch  = 0;
+    DataType x     = 0.;
+    DataType y     = 0.;
+    DataType r     = 0.;
+    DataType px    = 0.;
+    DataType py    = 0.;
+
+    DAAL_CHECK_MALLOC(posx);
```

Comment:
I don't think these checks are necessary because you've checked them before calling the function